HoodieCommitArchiveLog.openWriter causes object store to throw exceptions

Ran into following issue when trying to commit a batch into google cloud storage and it always fails for the first time and somehow powered through on the retry
```
com.uber.hoodie.exception.HoodieException: Unable to initialize HoodieLogFormat writer
    at com.uber.hoodie.io.HoodieCommitArchiveLog.openWriter(HoodieCommitArchiveLog.java:93)
    at com.uber.hoodie.io.HoodieCommitArchiveLog.archiveIfRequired(HoodieCommitArchiveLog.java:115)
    at com.uber.hoodie.HoodieWriteClient.commit(HoodieWriteClient.java:541)
    at com.uber.hoodie.HoodieWriteClient.commit(HoodieWriteClient.java:488)
    at com.uber.hoodie.HoodieWriteClient.commit(HoodieWriteClient.java:479)
    at com.uber.hoodie.HoodieSparkSqlWriter$.write(HoodieSparkSqlWriter.scala:162)
    at com.uber.hoodie.HoodieStreamingSink$$anonfun$1$$anonfun$apply$2.apply(HoodieStreamingSink.scala:51)
    at com.uber.hoodie.HoodieStreamingSink$$anonfun$1$$anonfun$apply$2.apply(HoodieStreamingSink.scala:51)
    at scala.util.Try$.apply(Try.scala:192)
    at com.uber.hoodie.HoodieStreamingSink$$anonfun$1.apply(HoodieStreamingSink.scala:50)
    at com.uber.hoodie.HoodieStreamingSink$$anonfun$1.apply(HoodieStreamingSink.scala:56)
    at com.uber.hoodie.HoodieStreamingSink.retry(HoodieStreamingSink.scala:99)
    at com.uber.hoodie.HoodieStreamingSink.addBatch(HoodieStreamingSink.scala:49)
    at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$org$apache$spark$sql$execution$streaming$MicroBatchExecution$$runBatch$3$$anonfun$apply$16.apply(MicroBatchExecution.scala:475)
    at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:77)
    at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$org$apache$spark$sql$execution$streaming$MicroBatchExecution$$runBatch$3.apply(MicroBatchExecution.scala:473)
    at org.apache.spark.sql.execution.streaming.ProgressReporter$class.reportTimeTaken(ProgressReporter.scala:271)
    at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:58)
    at org.apache.spark.sql.execution.streaming.MicroBatchExecution.org$apache$spark$sql$execution$streaming$MicroBatchExecution$$runBatch(MicroBatchExecution.scala:472)
    at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$runActivatedStream$1$$anonfun$apply$mcZ$sp$1.apply$mcV$sp(MicroBatchExecution.scala:133)
    at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$runActivatedStream$1$$anonfun$apply$mcZ$sp$1.apply(MicroBatchExecution.scala:121)
    at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$runActivatedStream$1$$anonfun$apply$mcZ$sp$1.apply(MicroBatchExecution.scala:121)
    at org.apache.spark.sql.execution.streaming.ProgressReporter$class.reportTimeTaken(ProgressReporter.scala:271)
    at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:58)
    at org.apache.spark.sql.execution.streaming.MicroBatchExecution$$anonfun$runActivatedStream$1.apply$mcZ$sp(MicroBatchExecution.scala:121)
    at org.apache.spark.sql.execution.streaming.ProcessingTimeExecutor.execute(TriggerExecutor.scala:56)
    at org.apache.spark.sql.execution.streaming.MicroBatchExecution.runActivatedStream(MicroBatchExecution.scala:117)
    at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runStream(StreamExecution.scala:279)
    at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:189)
Caused by: java.io.IOException: The append operation is not supported.
    at com.google.cloud.hadoop.fs.gcs.GoogleHadoopFileSystemBase.append(GoogleHadoopFileSystemBase.java:811)
    at org.apache.hadoop.fs.FileSystem.append(FileSystem.java:1176)
    at com.uber.hoodie.common.table.log.HoodieLogFormatWriter.<init>(HoodieLogFormatWriter.java:73)
    at com.uber.hoodie.common.table.log.HoodieLogFormat$WriterBuilder.build(HoodieLogFormat.java:204)
    at com.uber.hoodie.io.HoodieCommitArchiveLog.openWriter(HoodieCommitArchiveLog.java:88)
    ... 28 more
```

--- a/hoodie-client/src/main/java/com/uber/hoodie/io/storage/HoodieWrapperFileSystem.java
+++ b/hoodie-client/src/main/java/com/uber/hoodie/io/storage/HoodieWrapperFileSystem.java
@@ -16,16 +16,15 @@
 
 package com.uber.hoodie.io.storage;
 
+import com.uber.hoodie.common.storage.StorageSchemes;
 import com.uber.hoodie.common.util.FSUtils;
 import com.uber.hoodie.exception.HoodieIOException;
 import java.io.IOException;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.util.EnumSet;
-import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
-import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import org.apache.hadoop.conf.Configuration;
@@ -60,20 +59,6 @@ import org.apache.hadoop.util.Progressable;
 public class HoodieWrapperFileSystem extends FileSystem {
 
   public static final String HOODIE_SCHEME_PREFIX = "hoodie-";
-  private static final Set<String> SUPPORT_SCHEMES;
-
-  static {
-    SUPPORT_SCHEMES = new HashSet<>();
-    SUPPORT_SCHEMES.add("file");
-    SUPPORT_SCHEMES.add("hdfs");
-    SUPPORT_SCHEMES.add("s3");
-    SUPPORT_SCHEMES.add("s3a");
-
-    // Hoodie currently relies on underlying object store being fully
-    // consistent so only regional buckets should be used.
-    SUPPORT_SCHEMES.add("gs");
-    SUPPORT_SCHEMES.add("viewfs");
-  }
 
   private ConcurrentMap<String, SizeAwareFSDataOutputStream> openStreams = new
       ConcurrentHashMap<>();
@@ -104,7 +89,7 @@ public class HoodieWrapperFileSystem extends FileSystem {
 
   public static String getHoodieScheme(String scheme) {
     String newScheme;
-    if (SUPPORT_SCHEMES.contains(scheme)) {
+    if (StorageSchemes.isSchemeSupported(scheme)) {
       newScheme = HOODIE_SCHEME_PREFIX + scheme;
     } else {
       throw new IllegalArgumentException(

--- a/hoodie-common/src/main/java/com/uber/hoodie/common/storage/StorageSchemes.java
+++ b/hoodie-common/src/main/java/com/uber/hoodie/common/storage/StorageSchemes.java
@@ -1,0 +1,66 @@
+/*
+ * Copyright (c) 2018 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.hoodie.common.storage;
+
+import java.util.Arrays;
+
+/**
+ * All the supported storage schemes in Hoodie.
+ */
+public enum StorageSchemes {
+  // Local filesystem
+  FILE("file", false),
+  // Hadoop File System
+  HDFS("hdfs", true),
+  // Apache Ignite FS
+  IGNITE("igfs", true),
+  // AWS S3
+  S3A("s3a", false),
+  S3("s3", false),
+  // Google Cloud Storage
+  GCS("gs", false),
+  // View FS for federated setups. If federating across cloud stores, then append support is false
+  VIEWFS("viewfs", true);
+
+  private String scheme;
+  private boolean supportsAppend;
+
+  StorageSchemes(String scheme, boolean supportsAppend) {
+    this.scheme = scheme;
+    this.supportsAppend = supportsAppend;
+  }
+
+  public String getScheme() {
+    return scheme;
+  }
+
+  public boolean supportsAppend() {
+    return supportsAppend;
+  }
+
+  public static boolean isSchemeSupported(String scheme) {
+    return Arrays.stream(values()).filter(s -> s.getScheme().equals(scheme)).count() > 0;
+  }
+
+  public static boolean isAppendSupported(String scheme) {
+    if (!isSchemeSupported(scheme)) {
+      throw new IllegalArgumentException("Unsupported scheme :" + scheme);
+    }
+    return Arrays.stream(StorageSchemes.values())
+        .filter(s -> s.supportsAppend() && s.scheme.equals(scheme)).count() > 0;
+  }
+}

--- a/hoodie-common/src/main/java/com/uber/hoodie/common/table/log/HoodieLogFormatWriter.java
+++ b/hoodie-common/src/main/java/com/uber/hoodie/common/table/log/HoodieLogFormatWriter.java
@@ -17,6 +17,7 @@
 package com.uber.hoodie.common.table.log;
 
 import com.uber.hoodie.common.model.HoodieLogFile;
+import com.uber.hoodie.common.storage.StorageSchemes;
 import com.uber.hoodie.common.table.log.HoodieLogFormat.Writer;
 import com.uber.hoodie.common.table.log.HoodieLogFormat.WriterBuilder;
 import com.uber.hoodie.common.table.log.block.HoodieLogBlock;
@@ -68,26 +69,32 @@ public class HoodieLogFormatWriter implements HoodieLogFormat.Writer {
 
     Path path = logFile.getPath();
     if (fs.exists(path)) {
-      log.info(logFile + " exists. Appending to existing file");
-      try {
-        this.output = fs.append(path, bufferSize);
-      } catch (RemoteException e) {
-        log.warn("Remote Exception, attempting to handle or recover lease", e);
-        handleAppendExceptionOrRecoverLease(path, e);
-      } catch (IOException ioe) {
-        if (ioe.getMessage().equalsIgnoreCase("Not supported")) {
-          log.info("Append not supported. Opening a new log file..");
-          this.logFile = logFile.rollOver(fs);
-          createNewFile();
-        } else {
-          throw ioe;
+      boolean isAppendSupported = StorageSchemes.isAppendSupported(fs.getScheme());
+      if (isAppendSupported) {
+        log.info(logFile + " exists. Appending to existing file");
+        try {
+          this.output = fs.append(path, bufferSize);
+        } catch (RemoteException e) {
+          log.warn("Remote Exception, attempting to handle or recover lease", e);
+          handleAppendExceptionOrRecoverLease(path, e);
+        } catch (IOException ioe) {
+          if (ioe.getMessage().toLowerCase().contains("not supported")) {
+            // may still happen if scheme is viewfs.
+            isAppendSupported = false;
+          } else {
+            throw ioe;
+          }
         }
+      }
+      if (!isAppendSupported) {
+        this.logFile = logFile.rollOver(fs);
+        log.info("Append not supported.. Rolling over to " + logFile);
+        createNewFile();
       }
     } else {
       log.info(logFile + " does not exist. Create a new file");
       // Block size does not matter as we will always manually autoflush
       createNewFile();
-      // TODO - append a file level meta block
     }
   }
 

--- a/hoodie-common/src/test/java/com/uber/hoodie/common/storage/TestStorageSchemes.java
+++ b/hoodie-common/src/test/java/com/uber/hoodie/common/storage/TestStorageSchemes.java
@@ -1,0 +1,41 @@
+/*
+ * Copyright (c) 2018 Uber Technologies, Inc. (hoodie-dev-group@uber.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.hoodie.common.storage;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+import org.junit.Test;
+
+public class TestStorageSchemes {
+
+  @Test
+  public void testStorageSchemes() {
+    assertTrue(StorageSchemes.isSchemeSupported("hdfs"));
+    assertFalse(StorageSchemes.isSchemeSupported("s2"));
+    assertFalse(StorageSchemes.isAppendSupported("s3a"));
+    assertFalse(StorageSchemes.isAppendSupported("gs"));
+    assertTrue(StorageSchemes.isAppendSupported("viewfs"));
+    try {
+      StorageSchemes.isAppendSupported("s2");
+      fail("Should throw exception for unsupported schemes");
+    } catch (IllegalArgumentException ignore) {
+      // expected.
+    }
+  }
+}

